From c5e5e1c6f92bf7277aebdd1d0eddb9658b9c19c0 Mon Sep 17 00:00:00 2001
Message-Id: <c5e5e1c6f92bf7277aebdd1d0eddb9658b9c19c0.1373271642.git.jdenemar@redhat.com>
From: Gao feng <gaofeng@cn.fujitsu.com>
Date: Wed, 5 Jun 2013 16:37:59 -0600
Subject: [PATCH] libvirt: lxc: don't mkdir when selinux is disabled

https://bugzilla.redhat.com/show_bug.cgi?id=915485

libvirt lxc will fail to start when selinux is disabled.
error: Failed to start domain noroot
error: internal error guest failed to start: PATH=/bin:/sbin TERM=linux container=lxc-libvirt container_uuid=b9873916-3516-c199-8112-1592ff694a9e LIBVIRT_LXC_UUID=b9873916-3516-c199-8112-1592ff694a9e LIBVIRT_LXC_NAME=noroot /bin/sh
2013-01-09 11:04:05.384+0000: 1: info : libvirt version: 1.0.1
2013-01-09 11:04:05.384+0000: 1: error : lxcContainerMountBasicFS:546 : Failed to mkdir /sys/fs/selinux: No such file or directory
2013-01-09 11:04:05.384+0000: 7536: info : libvirt version: 1.0.1
2013-01-09 11:04:05.384+0000: 7536: error : virLXCControllerRun:1466 : error receiving signal from container: Input/output error
2013-01-09 11:04:05.404+0000: 7536: error : virCommandWait:2287 : internal error Child process (ip link del veth1) unexpected exit status 1: Cannot find device "veth1"

fix this problem by checking if selinuxfs is mounted
in host before we try to create dir /sys/fs/selinux.

Signed-off-by: Gao feng <gaofeng@cn.fujitsu.com>
(cherry picked from commit 8d63af22de880d34aa94bcaa7ed95a8eac856ac6)
---
 src/lxc/lxc_container.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/src/lxc/lxc_container.c b/src/lxc/lxc_container.c
index 59a2177..b33b0da 100644
--- a/src/lxc/lxc_container.c
+++ b/src/lxc/lxc_container.c
@@ -538,13 +538,6 @@ static int lxcContainerMountBasicFS(bool pivotRoot,
         VIR_DEBUG("Processing %s -> %s",
                   mnts[i].src, mnts[i].dst);
 
-        if (virFileMakePath(mnts[i].dst) < 0) {
-            virReportSystemError(errno,
-                                 _("Failed to mkdir %s"),
-                                 mnts[i].src);
-            goto cleanup;
-        }
-
         srcpath = mnts[i].src;
 
         /* Skip if mount doesn't exist in source */
@@ -552,6 +545,13 @@ static int lxcContainerMountBasicFS(bool pivotRoot,
             (access(srcpath, R_OK) < 0))
             continue;
 
+        if (virFileMakePath(mnts[i].dst) < 0) {
+            virReportSystemError(errno,
+                                 _("Failed to mkdir %s"),
+                                 mnts[i].src);
+            goto cleanup;
+        }
+
         VIR_DEBUG("Mount %s on %s type=%s flags=%x, opts=%s",
                   srcpath, mnts[i].dst, mnts[i].type, mnts[i].mflags, mnts[i].opts);
         if (mount(srcpath, mnts[i].dst, mnts[i].type, mnts[i].mflags, mnts[i].opts) < 0) {
-- 
1.8.2.1

